### PR TITLE
Feature/new creation wizard

### PR DESCRIPTION
**Pull request by andyjb**
# Critical Changes
New UI for creation of roll ups (not maintenance)

# Changes
From the initial PR, made some performance tuning changes, and also restricted the objects to those with a prefix, not least to get the drop down to render in the page

# Issues Closed

### CODE REVIEW

**Review by Riley-Schill — APPROVED** [review]
The new wizard looks great, I love not needing to pull the fields API name! 

Using the new wizard I tested the following: 
Deployed ContactTrigger and AccountTrigger 
Created a rollup to concatenate text and sum a number field from Contact to Account
Tested Relationship Criteria 
Created a custom object and deployed the trigger to that object, set contact as the parent and reran the above tests in the new wizard as well as the old one.

File: dlrs/main/classes/ManageLookupRollupSummariesNewController.cls [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+    public String selectedParentField {get;set;}
+    public String selectedChildField {get;set;}
+    public String selectedRField {get;set;}
+    public List<SelectOption> parentObjFields {get;set;}
+    public List<SelectOption> childObjectList {get;set;}
+    public List<SelectOption> childObjFields {get;set;}
+    public boolean isParentObjectSelected {get;set;}
+    public boolean isChildObjectSelected {get;set;}
+    //public Map<String,String> objectWithRelationshipField {get;set;}
+    public Map<String, List<String>> PnameWithRField {get;set;}
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewController.cls:19]: I'd prefer to remove commented code. 
This goes for all comments in this class. 
Since we haven't agreed on these code style questions, I won't reject the PR.

Comment (Szandor72) [reply dlrs/main/classes/ManageLookupRollupSummariesNewController.cls:19]: The rationale is the usual one: Code moves forward, comments don't. After a while the code in the comments becomes undeployable because the rest moved. 
it detracts from the actual code logic etc etc

File: dlrs/main/classes/ManageLookupRollupSummariesNewController.cls [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+                    ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.Error, recordError.Error)); 
+                }
+                for(String fieldError : recordError.FieldErrors) {
+                    ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.Error, fieldError));   
+                }
+            }
+        } catch (Exception e) {
+            string errormessage = e.getmessage();
+            if(selectedChildObject==null &&   errormessage.contains('Attempt to de-reference a null object')){
+                ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.Error, 'Please Select the Child Object'));
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewController.cls:112]: Aren't we using custom labels?

File: dlrs/main/classes/ManageLookupRollupSummariesNewController.cls [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+    public List<SelectOption> getCalculationSharingModes(){
+        List<SelectOption> options = new List<SelectOption>();
+        for(Schema.Picklistentry picklistEntry : LookupRollupSummary__c.fields.CalculationSharingMode__c.getDescribe().getPickListValues()) {
+            options.add(new SelectOption(picklistEntry.getValue(), picklistEntry.getLabel()));
+        }
+        return options;
+    }
+
+    public List<Selectoption> getParentObjList(){
+        objectDescription = Schema.getGlobalDescribe();
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewController.cls:169]: I think we talked about the fact that this can take a long time and return a huuuuuuge list

File: dlrs/main/classes/ManageLookupRollupSummariesNewController.cls [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+        List<SelectOption> objNames = new List<SelectOption>();
+        objNames.add(new SelectOption('','-- Select --'));
+        for(Schema.SObjectType obj : parentObjList)
+        {
+            Schema.DescribeSObjectResult description = obj.getDescribe(SObjectDescribeOptions.DEFERRED);
+            String prefix = description.getKeyPrefix();
+            
+            if (!String.isEmpty(prefix)) {
+                String objName = description.getName();
+                objNames.add(new SelectOption(objName, obj.getDescribe().getLabel()+' ('+ objName +')'));
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewController.cls:180]: Labels continued: getLabel() will return the localised or even overriden version of the object name. Hence we might have different languages being displayed in our picklists

File: dlrs/main/classes/ManageLookupRollupSummariesNewController.cls [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+        }
+        objNames.sort();
+        return objNames;
+    }
+
+    public void confirmParentObject(){
+        try{
+            if(selectedParentObject==null || selectedParentObject==''){
+                ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.Error, 'Please Select the Parent Object'));
+            }else{
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewController.cls:191]: a simple return after page messages would make `else` obsolete

File: dlrs/main/classes/ManageLookupRollupSummariesNewController.cls [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+                childObjectList = new List<SelectOption>();
+                parentObjFields = new List<SelectOption>();
+                ChildObjectsDuplicates = new Map<String,String>();
+                isParentObjectSelected = true;
+                childObjectList.add(new SelectOption('','-- Select --'));
+                parentObjFields.add(new SelectOption('','-- Select --')); 
+                Schema.SObjectType parentObject = objectDescription.get(selectedParentObject);
+                Schema.DescribeSObjectResult parentObjectDescribe = parentObject.getDescribe();
+                //AJB 18/11/21: changed to use a map to avoid string splitting and comparison later on
+                PnameWithRField = new Map<String, List<String>>();
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewController.cls:202]: Is this `parentNameWithRelationshipField`? Can we call it that way?

Comment (Szandor72) [reply dlrs/main/classes/ManageLookupRollupSummariesNewController.cls:202]: `parentNameToRelationshipFieldnames` ?

File: dlrs/main/classes/ManageLookupRollupSummariesNewController.cls [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+                }
+                for(String key : ChildObjectsDuplicates.keyset()){
+                    childObjectList.add(new SelectOption(key, ChildObjectsDuplicates.get(key)));
+                }
+                Map<String, schema.SObjectField> MapOfFields = parentObjectDescribe.fields.getMap();
+                for(Schema.SObjectField field : MapOfFields.values())
+                {
+                    Schema.DescribeFieldResult fr = field.getDescribe();
+                    //String dType = String.valueof(fr.getType());
+                    //if(dType == 'DOUBLE' || dType == 'INTEGER' || dType == 'CURRENCY' || dType == 'DATE' || dType == 'DATETIME' || dType == 'STRING'){
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewController.cls:236]: totally confusing, I am afraid. 

So initially only a basic list of fields was supported. now we seem to allow all field displaytypes - are we positive this works with time fields, geo location and what not?

File: dlrs/main/classes/ManageLookupRollupSummariesNewController.cls [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+                            for(String thisField : childRelationshipFields) {
+                                rFields.add(thisField);
+                            }
+                        }
+                    }
+
+                }
+                relationshipFields = new List<SelectOption>();
+                relationshipFields.add(new SelectOption('','-- Select --'));
+                if(rFields.size()>1){
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewController.cls:287]: use a switch statment

File: dlrs/main/classes/ManageLookupRollupSummariesNewController.cls [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+                    //}
+                // AJB 16/11/21: don't think we really need to loop over the map values twice
+                /*
+                } 
+
+                for(Schema.SobjectField strFld: mapOfFields.values())
+                {*/
+                    if(fr.getType() == Schema.DisplayType.REFERENCE)
+                    {
+                        if(String.valueOf(fr.getReferenceTo()) == '('+selectedParentObject+')'){
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewController.cls:276]: these two ifs can be merged into one

File: dlrs/main/classes/ManageLookupRollupSummariesNewController.cls-meta.xml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>52.0</apiVersion>
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewController.cls-meta.xml:3]: No Hidden API Upgrades. Package API Version is on 51.

File: dlrs/main/classes/ManageLookupRollupSummariesNewTest.cls [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        mrl.getParentObjList();
+        mrl.selectedParentObject = 'Account';
+        mrl.confirmParentObject();
+        mrl.selectedChildObject = 'Contact';
+        mrl.confirmChildObject();
+        mrl.updateChildField();
+        mrl.updateParentField();
+        mrl.updateRField();
+    }
+    static testMethod void myUnitTest2(){
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewTest.cls:46]: can we rename `myUnitTest1` and `myUnitTest2` and add asserts?

File: dlrs/main/classes/ManageLookupRollupSummariesNewTest.cls [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+
+        ManageLookupRollupSummariesNewController controller = new ManageLookupRollupSummariesNewController();
+
+        controller.LookupRollupSummary = rollupSummary;
+
+        Test.startTest();
+        PageReference nextPage = controller.save();
+        Test.stopTest();
+
+        System.assertEquals(true, nextPage.getUrl().startsWith('/apex/managelookuprollupsummaries'), 'The redirect went to the wrong place');
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewTest.cls:83]: Having a hard time wrapping my head around the assert message vs assert checks. 
`Did not receive expected return url after successful save`

File: dlrs/main/classes/ManageLookupRollupSummariesNewTest.cls [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+
+        controller.LookupRollupSummary = rollupSummary;
+
+        Test.startTest();
+        PageReference nextPage = controller.save();
+        Test.stopTest();
+
+        List<ApexPages.Message> messages = ApexPages.getMessages();
+        System.assertEquals(1, messages.size(), 'There should only be one error message');
+        System.assertEquals('Sum/Avg doesnt looks like valid for dates ! Still want, then implement the IRollerCoaster yourself and change this class as required.', messages[0].getSummary(), 'Unexpected error message received');
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewTest.cls:156]: This should be proof read as it is a user facing message

> Sum/Avg doesnt looks like valid for dates ! Still want, then implement the IRollerCoaster yourself and change this class as required.

File: dlrs/main/classes/ManageLookupRollupSummariesNewTest.cls-meta.xml [hunk 14]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>52.0</apiVersion>
```
Comment (Szandor72) [inline dlrs/main/classes/ManageLookupRollupSummariesNewTest.cls-meta.xml:3]: see above

File: dlrs/main/classes/MetadataServiceCalloutMock.cls-meta.xml [hunk 15]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>52.0</apiVersion>
```
Comment (Szandor72) [inline dlrs/main/classes/MetadataServiceCalloutMock.cls-meta.xml:3]: see above

File: dlrs/main/pages/managelookuprollupsummaries.page [hunk 16]
```diff
@@ -17,6 +17,7 @@
 				<apex:actionSupport event="onchange" action="{!load}" reRender="rollupDetailView" />
 				<apex:selectOptions value="{!Lookups}" />
 			</apex:selectList>
+			<apex:commandButton value="Try Our New Wizard" action="{!newWizard}" style="margin-left: 20px;" />
```
Comment (Szandor72) [inline dlrs/main/pages/managelookuprollupsummaries.page:20]: Custom Label?

File: dlrs/main/pages/managelookuprollupsummaries_New.page-meta.xml [hunk 17]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexPage xmlns="http://soap.sforce.com/2006/04/metadata"> 
+    <apiVersion>52.0</apiVersion>
```
Comment (Szandor72) [inline dlrs/main/pages/managelookuprollupsummaries_New.page-meta.xml:3]: see above

File: package-lock.json [hunk 18]
```diff
@@ -65,12 +65,30 @@
       }
     },
     "@babel/helper-annotate-as-pure": {
-      "version": "7.15.4",
-      "resolved": "https://registry.npmjs.org/@babel/helper-annotate-as-pure/-/helper-annotate-as-pure-7.15.4.tgz",
-      "integrity": "sha512-QwrtdNvUNsPCj2lfNQacsGSQvGX8ee1ttrBrcozUP2Sv/jylewBP/8QFe6ZkBsC8T/GYWonNAWJV4aRR9AL2DA==",
+      "version": "7.16.0",
```
Comment (Szandor72) [inline package-lock.json:68]: please revert these changes / do not commit package-lock.json

File: package.json [hunk 19]
```diff
@@ -18,19 +18,19 @@
     "precommit": "lint-staged"
   },
   "devDependencies": {
-    "@lwc/eslint-plugin-lwc": "^1.0.1",
+    "@lwc/eslint-plugin-lwc": "^1.1.1",
```
Comment (Szandor72) [inline package.json:21]: Please revert these changes / do not commit them

File: dlrs/main/pages/managelookuprollupsummaries_New.page [hunk 20]
```diff
@@ -0,0 +1,264 @@
+<apex:page controller="ManageLookupRollupSummariesNewController" lightningStylesheets="true" tabStyle="ManageLookupRollupSummaries__tab" showHeader="true" sidebar="true" action="{!init}">
```
Comment (Szandor72) [inline dlrs/main/pages/managelookuprollupsummaries_New.page:1]: Will we continue to have two setup pages? 
If yes, then please remove the `_` or rename to: ManageLookupRollupSummariesNew

reason: no underscores in metadata names

File: dlrs/main/pages/managelookuprollupsummaries_New.page [hunk 21]
```diff
@@ -0,0 +1,264 @@
...
+        <apex:actionFunction name="updateCalcMode" action="{!updateCalcMode}" reRender="refreshPCFields"/>
+        <apex:sectionHeader title="Manage Lookup Rollup Summaries"  subtitle="Manage, Deploy, Run and Schedule"/>
+        <apex:pageMessage rendered="{!$Setup.DeclarativeLookupRollupSummaries__c.HideManageLookupRollupSummariesInfo__c==false}" strength="1" escape="false" severity="Info" summary="This tab provides the ability to store rollup definitions as <b>Custom Metadata</b>. Allowing your rollups to be added to <b>Change Sets</b> and <b>Packages</b> and are also now automatically cloned during a <b>Sandbox</b> refresh. Note that you can also change configuration entered here under the <b>Setup</b> menu <b>Custom Metadata Types</b> page, however it is recommended you utilise this page as it provides added validation and features."><a href="{!URLFOR($Action.LookupRollupSummary__c.Tab,$ObjectType.LookupRollupSummary__c)}">It maybe that your org has rollups defined under the older Lookup Rollup Summaries tab here, it is recommended you move them over.</a>&nbsp;<apex:commandLink action="{!hideMessage}" value="Hide this message"/></apex:pageMessage>
```
Comment (Szandor72) [inline dlrs/main/pages/managelookuprollupsummaries_New.page:35]: Labels?

File: dlrs/main/pages/managelookuprollupsummaries_New.page [hunk 22]
```diff
@@ -0,0 +1,264 @@
...
+        <apex:pageMessages id="errorconfirm"/>
+        <apex:outputPanel rendered="{!MetadataConnectionError==true}">
+            <apex:outputLink value="{!$Page.welcometab}">Return to the Welcome page</apex:outputLink>
+        </apex:outputPanel>
+        <apex:outputPanel rendered="{!MetadataConnectionError==false}" id="refreshPb">             
+            <apex:pageBlock mode="edit" id="rollupDetailView" title="Lookup Rollup Summary Creation Wizard">
+                <apex:pageBlockButtons >
+
+                    <apex:commandButton value="Save" action="{!save}"/>
+                    <!--<apex:commandButton value="Delete" action="{!deleteX}" rendered="{!LookupRollupSummary.Id!=null}"/>
```
Comment (Szandor72) [inline dlrs/main/pages/managelookuprollupsummaries_New.page:45]: comments, again :)

**Review by Szandor72 — CHANGES_REQUESTED** [review]
At the very least, API Versions need to be corrected. 

We must make sure prettier works on your machine. If you like, I can do a final formatting when you're done. 

Thanks for your efforts!

**Review by Szandor72 — DISMISSED** [review]
As discussed with @afawcett today, we want to move this forward.
